docs/typescript.md

```diff
@@ -1,14 +1,170 @@
-# TypeScript Guidelines
+# **Typescript guidelines**
 
-These guidelines specifically apply to TypeScript.
+**_Purpose_**: Establish best practices for writing robust, maintainable, and efficient TypeScript code.
 
-## Provide explicit return types for functions/methods
+**_Audience_**: Internal and external developers using TypeScript in Metamask.
+
+## **Interfaces**
+
+By defining expected object structures with interfaces, TypeScript promotes developer intent and unlocks powerful IDE features like accurate completion and context-aware navigation.
+
+### **Enumerations**
+
+TypeScript offers several tools for crafting clear data definitions, with enumerations and unions standing as popular choices.
+
+**Consider using enums over union types for situations with a fixed set of known values.**
+
+Inevitably you will want to refer to the values of a union type somewhere (perhaps as the argument to a function). You can of course just use a literal which represents a member of that union — but if you have an enum, then all of the values are special, and any time you use a value then anyone can see where that value comes from.
+
+**Don't use numeric enums**
+
+Numeric enums are misleading because it creates a reverse mapping from value to property name, and when using `Object.values` to access member names, it will return the numerical values instead of the member names, potentially causing unexpected behavior.
+
+## **Types**
+
+#### Constrain generic types if necessary
+
+It may not be enough just to have a type or a function take another type — you might have to constrain it if it's not allowed to be anything (e.g. extends Json)
+
+```typescript
+// before
+function createExampleMiddleware(
+  exampleParam,
+): JsonRpcMiddleware<JsonRpcParams, Json>;
+// after
+function createExampleMiddleware<
+  Params extends JsonRpcParams = JsonRpcParmas,
+  Result extends Json = Json,
+>(exampleParam): JsonRpcMiddleware<Params, Result>;
+```
```

Comment:
This would be a much better example of generic constraints with the following change:

```suggestion
// before
function createExampleMiddleware<Params, Result>(
  exampleParam,
): JsonRpcMiddleware<JsonRpcParams, Json>;
// after
function createExampleMiddleware<
  Params extends JsonRpcParams,
  Result extends Json,
>(exampleParam): JsonRpcMiddleware<Params, Result>;
```
(removed the default params to keep the example focused on generic constraints and avoid confusion)

See: https://www.typescriptlang.org/docs/handbook/2/generics.html#generic-constraints